### ISSUE TITLE
Size of feature_names error, should equal with total number of features

### ISSUE THREAD

**Issue by weisskatze** [Microsoft/LightGBM#1887]
## Environment info

Operating System: Windows 10

CPU/GPU model: Ryzen 3 2200g

C++/Python/R version: R 3.5.1

## Error message

[LightGBM] [Fatal] Size of feature_names error, should equal with total number of features

## Reproducible examples

# get the numb 70/30 training test split
num_sam <- round(length(dt_label) * .7)

# training data
train_dt <- dt[1:num_sam,]
train_label <- dt_label[1:num_sam]

# testing data
test_data <- dt[-(1:num_sam),]
test_labels <- dt_label[-(1:num_sam),]

grid_search <- expand.grid(Depth = 2:3,
                           L2 = 0:3)

model <- list()
perf <- numeric(nrow(grid_search))

dtrain <- lgb.Dataset(train_dt, label = train_label, free_raw_data = FALSE)
dtest <- lgb.Dataset.create.valid(dtrain, test_data, label = test_labels)
valids <- list(test = dtest)

for (i in 1:nrow(grid_search)) {
  model[[i]] <- lgb.train(list(objective = "tweedie",
                               metric = "l2",
                               lambda_l2 = grid_search[i, "L2"],
                               max_depth = grid_search[i, "Depth"]),
                          dtrain,
                          2,
                          valids,
                          min_data = 1,
                          learning_rate = 1,
                          early_stopping_rounds = 5)
  perf[i] <- min(rbindlist(model[[i]]$record_evals$test$l2))
}

## Steps to reproduce

The features of test_set and train_set are totally equal. How can I deal with this issue?

@Laurae2
@BruceZhaoR
@guolinke
